### PR TITLE
[Arc] Add VectorizeOp canonicalization

### PR DESCRIPTION
**Pull request by elhewaty**

### CODE REVIEW

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (TaoBi22) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:624]: Nit, but will `uniqueOperands` ever not just be a copy of groups? As far as I can tell, the only case in which they would be different is where you break from the for loop over `group` early, in which case you'll return failure anyway as the size won't equal groupSize, so you could avoid a little bit of overhead with
```suggestion
    for (auto operand : group) {
      if (operand.getDefiningOp() != defOp)
        return failure();

      if (!operand.hasOneUse())
        return failure();
    }
```
or maybe this is just for type conversion?

Comment (elhewaty) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:624]: Thank you very much for the review.
> Will `uniqueOperands` ever not just be a copy of groups?

It will be a subset of `groups` of course but I need to make sure that all the operands of the definedOp appear in this operation and only have one use because this is one of the conditions to merge ops (This is an initial thoughts)

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (elhewaty) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:636]: @TaoBi22 @fabianschuiki @maerhart Can `RewriterBase::inlineRegionBefore` help with this?
or `RewriterBase::inlineBlockBefore` is better?

Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:636]: Yeah the inlining should be able to do this I believe. You might have to set up an `IRMapping` with mappings from the other vector's block args to the block args in the vector you're merging into.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (fabianschuiki) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:624]: You can probably very quickly check whether one of the vector inputs is defined by another vector by comparing the `OperandRange`s:
```
// This gives you an OperandRangeRange, i.e. a 2D array of operands.
// The outer dimension corresponds to arguments to the body block.
// The inner dimension corresponds to individual vector elements.
auto inputVecs = vecOp.getInputs();

// This iterates over the vectors flowing into the vectorize op:
for (auto inputVec : inputVecs) {
  // inputVec is an OperandRange, where each element corresponds to a vector element.
  // You can pick the first element, get its defining op, check if it's a vectorize op, and check if
  // all elements in `inputVec` are exactly the elements from that other vectorize op.
  auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
  if (!otherVecOp) continue;
  if (inputVec != otherVecOp.getResults()) continue;

  // Now you know that inputVec matches the results of otherVecOp exactly.
  // Now you can check if the current vecOp is the sole user of otherVecOp.
  if (!llvm::all_of(otherVecOp.getResults(), [](auto result) { return result.hasOneUse(); }))
    continue;

  // At this point you know that inputVec is exactly the result from otherVecOp, and
  // that this is the only use of otherVecOp. Now you should be able to merge the two.
}
```

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:630]: Using `rewriter.inlineBlockBefore` would avoid cloning operations since the original ops are removed anyways.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:645]: If the pattern cannot be applied, it should return `failure()`, but you need to make sure that no modification to the IR has already happened in that case.
Returning `success()` in such a case means that the operation is pushed onto the worklist again since the driver thinks the pattern was applied successfully and modified something. This leads to unnecessary inefficiencies.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (maerhart) [inline lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:652]: Nit:
```suggestion
  unsigned groupSize = vecOp.getResults().size();
  unsigned numOfGroups = newOperands.size() / groupSize;
  SmallVector<int32_t> newAttr(numOfGroups, groupSize);
```

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (maerhart) [inline test/Dialect/Arc/arc-canonicalizer.mlir:303]: Maybe add a test where multiple input vectors of a single `arc.vectorize` come from other vectorize ops and can be merged?

Comment (maerhart) [reply test/Dialect/Arc/arc-canonicalizer.mlir:303]: Also, it'd be nice to have a test for the case where the results of the vectorize op to be merged into the current one has other users.

Comment (maerhart) [reply test/Dialect/Arc/arc-canonicalizer.mlir:303]: And also one where the vector is not taken 1-1 but some reshuffling would need to happen.

**Review by maerhart — COMMENTED** [review]
Great job! So cool that this is working now.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (elhewaty) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:623]: I think there are no `insertArguments` in `mlir::Block` ? I think I will need to iterate over them for mapping ?

Comment (elhewaty) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:623]: I think `inlineBlockBefore` will handle this.

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (elhewaty) [reply test/Dialect/Arc/arc-canonicalizer.mlir:303]: > And also one where the vector is not taken 1-1 but some reshuffling would need to happen.

```
auto otherVecOp = inputVec[0].getDefiningOp<VectorizeOp>();
    auto resultVec = otherVecOp.getResults();
    auto comparator = [](const mlir::Value &a, const mlir::Value &b) {
      return a.getAsOpaquePointer() < b.getAsOpaquePointer();
    };
    std::sort(inputVec.begin(), inputVec.end(), comparator);
    std::sort(resultVec.begin(), resultVec.end(), comparator);
```

Can we sort both? if not do you have any other way?

Comment (elhewaty) [reply test/Dialect/Arc/arc-canonicalizer.mlir:303]: > And also one where the vector is not taken 1-1 but some reshuffling would need to happen.

Is there an easy way to sort `inputVec` and `resVec = otherVecOp.getResults()`?
Maybe I can store inputVec in `DensMap` and check but that will need to specialize the operators.

Comment (maerhart) [reply test/Dialect/Arc/arc-canonicalizer.mlir:303]: I don't understand why you would need to sort something here. For this test, your patten should not apply, i.e., it should not merge the two `arc.vectorize` ops.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (maerhart) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:630]: The body blocks don't have any successors since they are the only blocks in the region.
You'd need to modify the block arguments of the `otherVecOp` instead of the current block and pass the adequate values as the last argument of `inlineBlockBefore`. I don't think the code will get more complex by doing so, but it requires a few changes to your code. I'm also fine with leaving it as is for now. We can come back and get rid of the unnecessary cloning if we observe performance issues.

File: lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp
Comment (maerhart) [reply lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp:623]: Ah sry, I thought there exists a `insertArguments` because there also exists an `addArguments`, but you're right.

**Comment by fabianschuiki** [comment]
This is starting to look really nice! Any idea how well this works on the cores in the arc-tests repository? Are there any cases that don't work yet?

**Comment by elhewaty** [comment]
> This is starting to look really nice! Any idea how well this works on the cores in the arc-tests repository? Are there any cases that don't work yet?

It works well on the dual rocket core we vectorized before, but just a few merges!!
The problem here is that if we allow `ClockTypes` it crashes when at `otherArg.getType()` but after some work I noticed the problem is not that `ClockTypes` crashes but they may with some args that will make it crash as if we allow all args to be of the same type there's no crash happens!! so what can make `otherArg.getType` crash?!! that is the problem I can allow Clocks, Ints, and floats this may work with the tests but fail in some other cases.

**Review by fabianschuiki — APPROVED** [review]
Looks good to me! Nice job getting this up and running :slightly_smiling_face:.

I liked @maerhart's suggestion of having a test where one vectorize op feeds into another, but with the operands shuffled, such that the pattern should not apply. That might be worth adding before merging.

**Comment by elhewaty** [comment]
> I liked @maerhart's suggestion of having a test where one vectorize op feeds into another, but with the operands shuffled, such that the pattern should not apply. That might be worth adding before merging.

I think there is a test for this, there is a function called `Needs_Shuffle` at the enf of the tests? Or do you mean something else?

**Comment by fabianschuiki** [comment]
Oh you're totally right, I overlooked that. Very nice 😃

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (maerhart) [inline test/Dialect/Arc/arc-canonicalizer.mlir:406]: This could be another optimization pattern: if a vector is given multiple times as input to the vectorize op, we can remove one of them and replace usages with the other occurrence.
But that's something for another PR.

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (maerhart) [inline test/Dialect/Arc/arc-canonicalizer.mlir:420]: Nit: there are many unused input ports, we could trim them a bit.

**Review by maerhart — APPROVED** [review]
Nice work! 🎉

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (elhewaty) [reply test/Dialect/Arc/arc-canonicalizer.mlir:406]: I thought of this too. I will

File: test/Dialect/Arc/arc-canonicalizer.mlir
Comment (elhewaty) [reply test/Dialect/Arc/arc-canonicalizer.mlir]: Done! Please merge if it's ok

**Pull request merged**
